govcd/edgegateway.go

```diff
@@ -200,82 +200,84 @@ func (eGW *EdgeGateway) RemoveNATPortRule(networkHref, natType, externalIP, exte
 	}
 
-	newEdgeConfig := eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration
-
-	// Take care of the NAT service
-	newNatService := &types.NatService{}
-
-	newNatService.IsEnabled = newEdgeConfig.NatService.IsEnabled
-	newNatService.NatType = newEdgeConfig.NatService.NatType
-	newNatService.Policy = newEdgeConfig.NatService.Policy
-	newNatService.ExternalIP = newEdgeConfig.NatService.ExternalIP
-
-	for _, natRule := range newEdgeConfig.NatService.NatRule {
+	err := eGW.Refresh()
+	if err != nil {
+		return Task{}, fmt.Errorf("error refreshing edge gateway: %#v", err)
+	}
 
-		if natRule.RuleType == natType &&
-			natRule.GatewayNatRule.OriginalIP == externalIP &&
-			natRule.GatewayNatRule.OriginalPort == externalPort &&
-			natRule.GatewayNatRule.Interface.HREF == uplink.HREF {
-			util.Logger.Printf("[DEBUG] REMOVING %s Rule: %#v", natRule.RuleType, natRule.GatewayNatRule)
-			continue
+	natServiceToUpdate := eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService
+	ruleIndex := -1
+	if natServiceToUpdate != nil {
+		for n, existingNatRule := range natServiceToUpdate.NatRule {
+			if existingNatRule.ID == id {
+				ruleIndex = n
+				break
+			}
 		}
-		util.Logger.Printf("[DEBUG] KEEPING %s Rule: %#v", natRule.RuleType, natRule.GatewayNatRule)
-		newNatService.NatRule = append(newNatService.NatRule, natRule)
+	} else {
+		return Task{}, fmt.Errorf("edge gateway doesn't have NAT rules")
 	}
 
-	newEdgeConfig.NatService = newNatService
+	if ruleIndex == -1 {
+		return Task{}, fmt.Errorf("edge gateway doesn't have rule with such Id")
+	}
+
+	if len(natServiceToUpdate.NatRule) > 1 {
+		natServiceToUpdate.NatRule = append(natServiceToUpdate.NatRule[:ruleIndex], natServiceToUpdate.NatRule[ruleIndex+1:]...)
+	} else {
+		natServiceToUpdate.NatRule = nil
+	}
 
 	newRules := &types.EdgeGatewayServiceConfiguration{
 		Xmlns:      types.XMLNamespaceVCloud,
-		NatService: newNatService,
+		NatService: natServiceToUpdate,
 	}
 
-	apiEndpoint, _ := url.ParseRequestURI(eGW.EdgeGateway.HREF)
-	apiEndpoint.Path += "/action/configureServices"
+	egwConfigureHref, _ := url.ParseRequestURI(eGW.EdgeGateway.HREF)
+	egwConfigureHref.Path += "/action/configureServices"
 
 	// Return the task
-	return eGW.client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodPost,
+	return eGW.client.ExecuteTaskRequest(egwConfigureHref.String(), http.MethodPost,
 		"application/vnd.vmware.admin.edgeGatewayServiceConfiguration+xml", "error reconfiguring Edge Gateway: %s", newRules)
-
 }
 
-// AddDNATRule creates DNAT rule and return refreshed existing NAT rules array or error
-// Allows to assign specific network Org VDC or external. Old function AddNATPortMapping and
-// AddNATMapping assigns rule to first external network
-func (eGW *EdgeGateway) AddDNATRule(ruleDetails NatRule) ([]*types.NatRule, error) {
+// AddDNATRule creates DNAT rule and return created NAT struct or error.
+// Allows to assign specific Org VDC or external network.
```

Comment:
"Allows to assign specific Org VDC or external network" -> "Allows assigning a specific Org VDC or an external network"

Reply:
Changed